src/node_api_types.h

```diff
@@ -0,0 +1,95 @@
+#ifndef SRC_NODE_API_TYPES_H_
+#define SRC_NODE_API_TYPES_H_
+
+#include <stddef.h>
+#include <stdint.h>
+
+#if !defined __cplusplus || (defined(_MSC_VER) && _MSC_VER < 1900)
+    typedef uint16_t char16_t;
+#endif
+
+// JSVM API types are all opaque pointers for ABI stability
+// typedef undefined structs instead of void* for compile time type safety
+typedef struct napi_env__ *napi_env;
+typedef struct napi_value__ *napi_value;
+typedef struct napi_ref__ *napi_ref;
+typedef struct napi_handle_scope__ *napi_handle_scope;
+typedef struct napi_escapable_handle_scope__ *napi_escapable_handle_scope;
+typedef struct napi_callback_info__ *napi_callback_info;
+
+typedef void (*napi_callback)(napi_env env,
+                              napi_callback_info info);
+typedef void (*napi_finalize)(napi_env env,
+                              void* finalize_data,
+                              void* finalize_hint);
```

Comment:
> `NAPI_EXTERN` is not appropriate here because typedefs are not symbols that get exported or imported.

That’ true, but you also have that in `node_api.h`. I’d remove it in both places then.

> I thought `__cdecl` would be the default for both C and C++, but maybe it is different for some platforms/compilers that node supports?

Mh, I think the x64 default calling conventions don’t match __cdecl. It’s been a while since I’ve had to deal with this stuff… leaving it this way should be fine.